### ISSUE TITLE
Teachers not showing

### ISSUE THREAD

**Issue by JustABlitz** [SapuSeven/BetterUntis#328]
I really dont know how to write this, but basically I have access to teachers on normal Untis but not on BetterUntis. Everyone else in my class also has access to teachers (they use normal untis) so it has to be a betteruntis issue I think

**Comment by SapuSeven** [comment]
Just to confirm - the teacher list when selecting teachers in BetterUntis is completely empty?

**Comment by AdditionalRAM** [comment]
I have a similar issue. The teacher list is completely empty, and no teachers are shown on the lesson info screen which is not the case on proprietary Untis.

**Comment by AdditionalRAM** [comment]
Here are some screenshots, if it helps. In this case, the teacher's name is "Zei", which is short for "Zeitler"
![teachers](https://user-images.githubusercontent.com/44118409/210222864-eb14e110-242a-451e-b5f7-e4a9280e4c88.jpeg)
![betteruntis-lessoninfo](https://user-images.githubusercontent.com/44118409/210222866-585db628-8831-403d-a101-31237f365e59.jpeg)
![untis-lessoninfo](https://user-images.githubusercontent.com/44118409/210222868-f71eccb7-d709-419a-92c1-e8e6ae69b7db.jpeg)

**Comment by lm41** [comment]
Is this still an issue?

**Comment by AdditionalRAM** [comment]
Unfortunately, yes. Even on the latest beta version, it is still a problem for me.

**Comment by Greyhash-dev** [comment]
Same issue here, teachers are not showing (even on the newest build)

**Comment by Greyhash-dev** [comment]
It's a kinda big issue for me because it does not show when lessions are held by a different (irregular) teacher (you can see what I mean by looking at the attached screenshots, look at the "current" lession for an example)
![Screenshot_20230925-111903_BetterUntis~2](https://github.com/SapuSeven/BetterUntis/assets/65809634/ef15bc7f-1c67-407d-8c72-b978deecd46f)

![Screenshot_20230925-111914_Untis Mobile~2](https://github.com/SapuSeven/BetterUntis/assets/65809634/b1e04ffc-459f-42e4-aeb3-d9f54056817d)

**Comment by lm41** [comment]
![Screenshot_20230925-163452~3.png](https://github.com/SapuSeven/BetterUntis/assets/59062169/4018b57d-5232-4957-811c-8351acac5d48)



Is this setting turned on? If not can you turn it on and check if this fixes the issue that lessons with another teacher aren't displayed correctly?

**Comment by AdditionalRAM** [comment]
Yes, it is turned on. I have the same issue, I can't see teachers no matter what. I do have the setting turned on. 
Whenever I have a substitution, I have to open the normal Untis app or the WebUntis page in order to check. 

![Screenshot_20230925_165901_com grupet web app](https://github.com/SapuSeven/BetterUntis/assets/44118409/52b9a3f9-63ca-4149-a7c0-cb60e4faaeab)
![Screenshot_20230925_165842_com sapuseven untis](https://github.com/SapuSeven/BetterUntis/assets/44118409/41a5b169-d881-47ec-8b4a-795a8477977b)
![Screenshot_20230925_165834_com sapuseven untis](https://github.com/SapuSeven/BetterUntis/assets/44118409/b4393b0e-2423-4af3-9c48-6a144e00104b)

**Comment by Greyhash-dev** [comment]
Yes, I've got that enabled too
![Screenshot_20230925-195821_BetterUntis](https://github.com/SapuSeven/BetterUntis/assets/65809634/e5fef355-474e-4bef-971f-26682215a8f0)

**Comment by SapuSeven** [comment]
Did you try re-logging into your profile?
Might be a long shot, but worth a try

**Comment by Greyhash-dev** [comment]
Yes, tried that twice, just to be sure, but unfortunately still the same /:

**Comment by AdditionalRAM** [comment]
Yeah. I had to configure multiple devices multiple times, including my Ticwatch E2. No luck with the beta versions either.

**Comment by RedCommander735** [comment]
Hi, as mentioned in #158 the teacher names won’t show up for me either. I was playing around with the WebUntis API (JSONRPC 2.0) myself, and found out that at least for my school, the API straight up will not return anything concerning teachers. The getTeachers method returns an empty array, and even if I specifically use the customizable version of the getTimetable method, and ask for teacher IDs and names (`{'options': {
        "element": {"id": person_id, "type": 5},  
        "startDate": 20231006,  
        "endDate": 20231006,  
        "teacherFields": ["id", "name"]  
}}`) the response object will not have teacher fields (`[{'id': 2352856, 'date': 20231006, 'startTime': 1030, 'endTime': 1115, 'kl': [{'id': 1256}], 'su': [{'id': 250}], 'ro': [{'id': 133}], 'activityType': 'Lessons'}, ...]`). It seems they have the option to disable access to teachers for the api?
  
The website seems to use some other api.

**Comment by AdditionalRAM** [comment]
That is interesting. I will email our head teacher (who manages the timetable and Untis) and will update here if/when I get a response.

**Comment by AdditionalRAM** [comment]
Mr Stendel invited me to his office and showed me around the UI and settings menu. Understandably, he couldn't give me any passwords or similar, but he helped me the best he could. We unfortunately couldn't find anything that resembles API settings or data sharing with 3rd parties or similar. A lot of you probably know this already, but I also learned:
-Untis is a local program where changes to the substitution table or timetable can be made.
-These changes are then pushed to the WebUntis servers.
-The Untis Mobile app and the API receive data from the WebUntis servers.
We couldn't find anything, neither in the Untis settings nor in WebUntis. We also checked the website for any potential documentation on this with no luck.
Even though we couldn't find such an option, thanks to Mr Stendel for the help!

**Comment by RedCommander735** [comment]
So I was curious and wrote an email to Untis support. They told me that the current JSON RPC API will be discontinued at the end of the year. They said the new API wasn't finished yet and there is no documentation yet, and I should contact them again at the end of the term, so they could hopefully provide me with more information. The new API will not be compatible though.

> "die Json-API wird voraussichtlich mit Ende des Jahres eingestellt. Es gibt weder Support, noch Fehlerbehebungen/Features mehr dafür. Die neue, nicht kompatible, auf Basis der Plattform laufenden API, ist aber noch nicht fertig und es gibt auch noch keine Dokumentation oder genauen Veröffentlichungstermin dafür. Ich bitte Sie deshalb sich zum Ende des ersten Semesters noch einmal zu melden, dann kann ich Ihnen hoffentlich mehr dazu mitteilen."  
  
That probably will take a bit of work to migrate to a new API.

**Comment by SapuSeven** [comment]
Thank you all for your efforts.
Right now I can't reproduce this issue myself and without network traces of the original app it's impossible to find the reason for the missing teachers, so I can't work on this issue.
I hope there's more success when their new API is released - it looks like Untis is even willing to provide some documentation for it, which would save a huge amount of time and effort.

**Comment by SapuSeven** [comment]
Please check out the latest debug build (easiest to download [from my website](https://api.sapuseven.com/apps/betteruntis/artifact/debug/1269)) - I added the option to view raw data of timetable items.
Just open the details screen of any affected item and press the bug-icon in the top right corner.
Paste the raw data here along with a screenshot of the regular details dialog, I want to check if the teachers are even sent at all.

**Comment by AdditionalRAM** [comment]
```{
  "element": {
    "id": 1589200,
    "lessonId": 15507,
    "startDateTime": "2023-11-06T08:00Z",
    "endDateTime": "2023-11-06T08:45Z",
    "foreColor": "#000000",
    "backColor": "#f49f25",
    "innerForeColor": "#000000",
    "innerBackColor": "#000000",
    "text": {
      "lesson": "",
      "substitution": "",
      "info": "",
      "staffInfo": null,
      "attachments": [
      ],
      "staffAttachments": null
    },
    "elements": [
      {
        "type": "CLASS",
        "id": 527,
        "orgId": 527
      },
      {
        "type": "SUBJECT",
        "id": 245,
        "orgId": 245
      },
      {
        "type": "ROOM",
        "id": 367,
        "orgId": 367
      }
    ],
    "can": [
      "READ_PERIODINFO"
    ],
    "is": [
      "REGULAR"
    ],
    "homeWorks": null,
    "exam": null,
    "isOnlinePeriod": false,
    "onlinePeriodLink": null,
    "messengerChannel": null,
    "blockHash": -75981911
  },
  "classes": [
    {
      "type": "CLASS",
      "id": 527,
      "orgId": 527
    }
  ],
  "teachers": [
  ],
  "subjects": [
    {
      "type": "SUBJECT",
      "id": 245,
      "orgId": 245
    }
  ],
  "rooms": [
    {
      "type": "ROOM",
      "id": 367,
      "orgId": 367
    }
  ],
  "forceIrregular": false
}
```

Teachers is just completely empty, as a user already said. Here's a screenshot:
![Screenshot_20231105_100317_com sapuseven untis debug](https://github.com/SapuSeven/BetterUntis/assets/44118409/e7d390e8-a066-4536-8663-c3c32e91e274)

**Comment by RedCommander735** [comment]
Teachers are empty :(

<details>
<summary>Raw Data</summary>

```
{
  "element": {
    "id": 2411488,
    "lessonId": 84297,
    "startDateTime": "2023-11-06T07:50Z",
    "endDateTime": "2023-11-06T09:25Z",
    "foreColor": "#000000",
    "backColor": "#99cccc",
    "innerForeColor": "#000000",
    "innerBackColor": "#000000",
    "text": {
      "lesson": "A24L3_1, A24L3",
      "substitution": "",
      "info": "",
      "staffInfo": null,
      "attachments": [
      ],
      "staffAttachments": null
    },
    "elements": [
      {
        "type": "CLASS",
        "id": 1256,
        "orgId": 1256
      },
      {
        "type": "SUBJECT",
        "id": 311,
        "orgId": 311
      },
      {
        "type": "ROOM",
        "id": 56,
        "orgId": 56
      }
    ],
    "can": [
      "READ_HOMEWORK",
      "READ_PERIODINFO"
    ],
    "is": [
      "REGULAR"
    ],
    "homeWorks": [
    ],
    "exam": null,
    "isOnlinePeriod": false,
    "onlinePeriodLink": null,
    "messengerChannel": null,
    "blockHash": 2041249027
  },
  "classes": [
    {
      "type": "CLASS",
      "id": 1256,
      "orgId": 1256
    }
  ],
  "teachers": [
  ],
  "subjects": [
    {
      "type": "SUBJECT",
      "id": 311,
      "orgId": 311
    }
  ],
  "rooms": [
    {
      "type": "ROOM",
      "id": 56,
      "orgId": 56
    }
  ],
  "forceIrregular": false
}
```

</details>
<details>
<summary>Image</summary>
<img src='https://github.com/SapuSeven/BetterUntis/assets/40245965/1dfca7ed-6106-4389-afcd-43d88c8a979d' style="height: 50px;">
</details>

**Comment by SapuSeven** [comment]
That's unfortunate.
I think we need to wait for the new API to be ready then.

**Comment by AdditionalRAM** [comment]
Any updates on this?

**Comment by RedCommander735** [comment]
> leider kann ich Ihnen auch jetzt keine neuen Informationen zur neuen API geben. Aus verschiedenen Gründen (Ressourcen, Vorziehungen usw.) verzögert sich die Fertigstellung immer wieder. Ich übermittle Ihnen anbei die alte JSON-API Dokumentation, da es derzeit nicht absehbar ist, wann die neue API kommen wird. Definitiv gibt es aber keinen Support mehr für die JSON-API. Nachfragen dazu sind also sinnlos.

&rarr; No new API in the near future. 
> For various reasons (resources, advances, etc.) completion is repeatedly delayed.

**Comment by Jakobwoi** [comment]
As I have the same problem I also wrote an e-mail to the untis support.
> Unsere neue, öffentliche API ist leider immer noch nicht fertig. Ich rechne damit frühestens zu Semesterende bzw. vermutlich erst zum Start des neuen Schuljahres (~September). Die JSON-RPC API wird natürlich mindestens so lange weiter verwendbar sein.

→So new API, at the earliest in september.

> Our new, public API is sadly still not finished. I expect that the new API will be available at the earliest at the end of the semester or the beginning of the new school year. The JSON-RPC will of course be available at least until then.

**Comment by SapuSeven** [comment]
Interesting.
I hope that the "public API" will actually be public and have the same capabilities as they use internally.

**Comment by unatried** [comment]
@SapuSeven are you already using this API?
https://developer.untis.com/core-concepts/authentication-model/

**Comment by michilyy** [comment]
Hello, are there any news on this issue?
